specs/PushNotifications/ToastNotifications-spec.md

```diff
@@ -0,0 +1,940 @@
+        ToastNotificationManager::ToastActivated(token);
+    }
+    else if (kind == ExtendedActivationKind::ToastNotification)
+    {
+        auto toastActivatedEventArgs = args.Data().as<ToastActivatedEventArgs>();
+        ProcessToastArgs(toastActivatedEventArgs);
+    }
+
+    return 0;
+}
+```
+
+## Unpackaged app Toast Registration
+
+Similarly unpackaged apps can also Register themselves to send and act as targets for toast notifications.
+
+```cpp
+int main()
+{
+    ToastAssets assets = winrt::make<ToastAssets>(
+        L"SampleApp",
+        L"c:\test\icon.png",
+        Windows::UI::Colors::LightGray());
+    auto activationInfo = ToastActivationInfo::CreateToastActivationInfoFromToastAssets(assets);
+    ToastNotificationManager::RegisterActivator(activationInfo);
+
+    auto args = AppInstance::GetCurrent().GetActivatedEventArgs();
+    auto kind = args.Kind();
+
+    if (kind == ExtendedActivationKind::Launch)
+    {
+        // App is launched in FG. So intercept toast activators via FG event
+        const auto token = ToastNotificationManager::ToastActivated([](const auto&, ToastActivatedEventArgs const& toastActivatedEventArgs)
+            {
+                ProcessToastArgs(toastActivatedEventArgs);
+            });
+
+        // App does Foreground Stuff Here
+
+        // Cleanup
+        ToastNotificationManager::ToastActivated(token);
+    }
+    else if (kind == ExtendedActivationKind::ToastNotification)
+    {
+        auto toastActivatedEventArgs = args.Data().as<ToastActivatedEventArgs>();
+        ProcessToastArgs(toastActivatedEventArgs);
+    }
+
+    return 0;
+}
+```
+
+## Displaying a Toast
+
+The app will need to display a toast to the user. To do this we recommend that the developer always sets a tag and group field so that similar toasts can replace it in the future. They also have the option to delete the toast automatically from ActionCentre on subsequent reboots by setting the ExpiresOnReboot property to true.
+
+```cpp
+// Display a toast with tag and group
+void DisplayToast(const winrt::Windows::Data::Xml::Dom::XmlDocument& doc)
+{
+    ToastNotification toast = winrt::make<ToastNotification>(doc);
+    // Setting Tag, Group and ExpiresOnReboot properties are optional
+    toast.Tag(L"Tag");
+    toast.Group(L"Group");
+    toast.ExpiresOnReboot(true);
+    ToastNotificationManager::ShowToast(toast);
+}
+```
+
+## Create a Toast Payload using Preexisting Toast XML templates
+
+We will provide a set of templates that developers can use to compose payloads. The developer can then use the Document Object Model (DOM) to customize the content in the payload. The templates that we are introducing are as follows:
+
+| Template Name | Scenario Examples | Example |
+| :--- | :---| :--- |
+|BasicToast| Simple alerts | ![BasicToast](images/BasicToast.png) |
+|ToastWithTwoButtons| Quick actions (2 options) such as view more information or remind me later | ![BasicToast](images/ToastWithTwoButtons.png) |
```

Comment:
Should we have a toast with one button? Or do we think thats not critical?

Reply:
@vaheeshta

Reply:
Shell believes not critical, but toast with icons is, so added that.

Reply:
Thanks @vaheeshta .